apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,328 @@
+                        .map_err(|e| Status::internal(e.to_string()))?;
+                    tracing::debug!("ingress response text: {:?}", data);
+                    let response = ReporterResponse {
+                        message: "Report accepted".to_string(),
+                    };
+                    return Ok(Response::new(response));
+                }
+                Err(e) => {
+                    tracing::warn!("attempt: {}, could not transfer: {}", i + 1, e);
+                    if i == 4 {
+                        return Err(Status::unavailable(e.to_string()));
+                    }
+                    backoff += Duration::from_millis(50);
+                    tokio::time::sleep(backoff).await;
+                }
+            }
+        }
+        // The compiler can't figure out the exit paths are covered,
+        // so to keep it happy have...
+        Err(Status::unavailable("should not happen..."))
+    }
+}
+
+#[tonic::async_trait]
+impl Reporter for ReportSpaceport {
+    async fn add_stats(
+        &self,
+        request: Request<ReporterStats>,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        tracing::debug!("received request: {:?}", request);
+        let msg = request.into_inner();
+        self.add_stats_or_trace(StatsOrTrace::Stats(msg)).await
+    }
+
+    async fn add_trace(
+        &self,
+        request: Request<ReporterTrace>,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        tracing::debug!("received request: {:?}", request);
+        let msg = request.into_inner();
+        self.add_stats_or_trace(StatsOrTrace::Trace(msg)).await
+    }
+}
+
+impl ReportSpaceport {
+    async fn add_stats_or_trace(
+        &self,
+        record: StatsOrTrace,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        let mut graph_usage = self.graph_usage.lock().await;
+        let graph_map = graph_usage
+            .entry(record.graph().unwrap())
+            .or_insert_with(HashMap::new);
+        let entry = graph_map
+            .entry(record.key())
+            .or_insert_with(|| record.get_traces_and_stats());
+        match record {
+            StatsOrTrace::Stats(mut s) => entry.stats_with_context.push(s.stats.take().unwrap()),
+            StatsOrTrace::Trace(mut t) => entry.trace.push(t.trace.take().unwrap()),
+        }
+        // Drop the graph_usage lock to maximise concurrency
+        drop(graph_usage);
+
+        // This is inherently both imprecise and racy, but it doesn't matter
+        // because we are just hinting to the spaceport that it's probably a
+        // good idea to try to transfer data up to the ingress. Multiple
+        // notifications just trigger more transfers.
+        //
+        // TRIGGER_BATCH_LIMIT is a fairly arbitrary number which indicates
+        // that we are adding a lot of data for transfer. It is derived
+        // empirically from load testing.
+        let total = self.total.fetch_add(1, Ordering::SeqCst);
```

Comment:
I'm not sure I follow what this is counting. Overall operations (requests) processed? Given that for now the only thing you're tracking per operation is a duration histogram (no traces yet, right?), I'd suggest that this should be counting the cardinality of unique operations.  Having 50 (or whatever) different operations could trigger the size limit but 50 requests on the same operation shouldn't take much more space than 1.

Reply:
This number is trying to deal with the more complex scenario of multiple routers all sending data to a single, shared spaceport. It's certainly not an exact science right now (hence the comment about empirical derivation). I think this is something that we can improve over time and with experience from usage. I'll file an issue which tries to capture some of this.

Reply:
sure, my point is that the concept of "send reports earlier than the regularly scheduled interval" should probably be based on the estimated size of the report, and "number of unique operations" is going to dominate a lot more than "number of operations", since we're aggregating stats!